infoschema/infoschema.go

```diff
@@ -403,3 +403,129 @@ func GetBundle(h InfoSchema, ids []int64) *placement.Bundle {
 	}
 	return &placement.Bundle{ID: placement.GroupID(id), Rules: newRules}
 }
+
+// LocalTemporaryTableInfoSchema is the interface used to retrieve the local temporary table schema.
+type LocalTemporaryTableInfoSchema interface {
+	TableByName(schema, table model.CIStr) (*LocalTemporaryTable, error)
+	TableExists(schema, table model.CIStr) bool
+	TableByID(id int64) (*LocalTemporaryTable, bool)
+	AddTable(schema *model.DBInfo, table table.Table) error
+	RemoveTable(schema, table model.CIStr) bool
+}
+
+// LocalTemporaryTable stores table and schema info for local temporary table.
+// Because drop database doesn't drop its local temporary table. So each table's database may be different event it has the same name.
+type LocalTemporaryTable struct {
+	table.Table
+	Schema *model.DBInfo
+}
+
+type localTempSchemaTables struct {
+	tables map[string]*LocalTemporaryTable
+}
+
+// localTemporaryTableInfoSchema implements LocalTemporaryTableInfoSchema
+type localTemporaryTableInfoSchema struct {
+	schemaMap map[string]*localTempSchemaTables
+	idx2table map[int64]*LocalTemporaryTable
+}
```

Comment:
Any reasons why not reuse `infoSchema` and `schemaTables`?

Reply:
If we drop a database and then create a now one with a same name in a single session, it can happens that two temporary table belong to different databases but with the same name. Though it works well when using schemaTables because db information for local temporary table is not necessary, it is clear to present the real situation to store a dbinfo in each table.

Another reason why not using infoSchema is that it is immutable and has many method we do not need, it more simple to define a new structure.

Reply:
Define another `LocalTemporaryTableInfoSchema` interface is unnecessary.
It's easy to override the `InfoSchema` interface providing a different implementation.

https://github.com/pingcap/tidb/pull/22440/files#diff-a40af26d74c0e0d741ec31becd7338558785033cf251e8ef1c61d1eed7a4ed31R8

Reply:
@tiancaiamao I'm ok to remove interface LocalTemporaryTableInfoSchema and override InfoSchema. But because session use function GetInfoSchema to get a final infoschema instead of just getting it from TxnCtx, I should rethink how to store it in the sessionvars